File: playlists.py
"""
Spotify playlists — list mode.
"""

import sys

from lib import (
    read_input,
    write_response,
    list_response,
    error,
    need_setup_response,
    is_authed,
    api,
)


def main():
    inp = read_input()
    ctx = inp.get("context", {})

    if not is_authed(ctx):
        write_response(need_setup_response())
        return

    try:
        data = api(ctx, "GET", "/me/playlists", params={"limit": 50})
    except RuntimeError as e:
        if str(e) == "not_authenticated":
            write_response(need_setup_response())
            return
        write_response(error("Playlists failed", details=str(e)))
        return

    items = (data or {}).get("items") or []
    if not items:
        write_response(
            list_response(
                [
                    {
                        "id": "empty",
                        "title": "No playlists",
                        "subtitle": "Create one in Spotify and reload.",
                        "icon": "L",
                    }
                ]
            )
        )
        return

    out = []
    for p in items:
        owner = (p.get("owner") or {}).get("display_name", "")
        tracks = (p.get("tracks") or {}).get("total", 0)
        subtitle_parts = []
        if owner:
            subtitle_parts.append(f"by {owner}")
        subtitle_parts.append(f"{tracks} tracks")
        if p.get("public") is False:
            subtitle_parts.append("private")
        out.append(
            {
                "id": p.get("uri", p.get("id", "")),
                "title": p.get("name", "Untitled"),
                "subtitle": "  ·  ".join(subtitle_parts),
                "icon": "L",
                "actions": [
                    {
                        "title": "Play",
                        "shortcut": "enter",
                        "type": "exec",
                        "exec": "playlists.py",
                        "args": ["play", p.get("uri", "")],
                    },
                    {
                        "title": "Open in Spotify",
                        "shortcut": "cmd+o",
                        "type": "open_url",
                        "url": (p.get("external_urls") or {}).get("spotify", ""),
                    },
                ],
            }
        )

    write_response(list_response(out))


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        write_response(error("playlists crashed: " + str(e)))
        sys.exit(1)

File: playlists.js
#!/usr/bin/env node
"use strict";

/**
 * Spotify playlists — list mode.
 */

const {
  readInput,
  writeResponse,
  list,
  error,
  needSetupResponse,
  isAuthed,
  api,
} = require("./lib");

async function main() {
  const inp = await readInput();
  const ctx = inp.context || {};

  if (!isAuthed(ctx)) {
    writeResponse(needSetupResponse());
    return;
  }

  let data;
  try {
    data = await api(ctx, "GET", "/me/playlists", null, { limit: 50 });
  } catch (e) {
    if (e.message === "not_authenticated") {
      writeResponse(needSetupResponse());
      return;
    }
    writeResponse(error("Playlists failed", e.message));
    return;
  }

  const items = (data || {}).items || [];
  if (!items.length) {
    writeResponse(
      list([
        {
          id: "empty",
          title: "No playlists",
          subtitle: "Create one in Spotify and reload.",
          icon: "L",
        },
      ])
    );
    return;
  }

  const out = items.map(function (p) {
    const owner = (p.owner || {}).display_name || "";
    const tracks = (p.tracks || {}).total || 0;
    const subtitleParts = [];
    if (owner) subtitleParts.push("by " + owner);
    subtitleParts.push(tracks + " tracks");
    if (p.public === false) subtitleParts.push("private");

    return {
      id: p.uri || p.id || "",
      title: p.name || "Untitled",
      subtitle: subtitleParts.join("  \u00B7  "),
      icon: "L",
      actions: [
        {
          title: "Play",
          shortcut: "enter",
          type: "exec",
          exec: "playlists.js",
          args: ["play", p.uri || ""],
        },
        {
          title: "Open in Spotify",
          shortcut: "cmd+o",
          type: "open_url",
          url: (p.external_urls || {}).spotify || "",
        },
      ],
    };
  });

  writeResponse(list(out));
}

main().catch(function (e) {
  writeResponse(error("playlists crashed: " + e.message));
  process.exit(1);
});
